src/org/gk/pro/ProExporter.java

```diff
@@ -0,0 +1,643 @@
+            modifications = residueExporter.exportModification((GKInstance) modifiedResidue);
+            if (modifications != null)
+                output += modifications;
+        }
+        resetIndices();
+
+        return output;
+    }
+
+    /**
+     * Return the modifications for a given IntraChainCrosslinkedResidue.
+     *
+     * Special case of {@link #getModifications(GKInstance, List)}
+     *
+     * @param modifiedResidue
+     * @return String
+     * @throws Exception
+     */
+    private String getIntraChainModifications(List<Object> modifiedResidues) throws Exception {
+        String output = "";
+        String modification = null;
+        IntraChainCrosslinkedResidue residueExporter = new IntraChainCrosslinkedResidue();
+        for (Object modifiedResidue : modifiedResidues) {
+            modification = residueExporter.exportPsiModIdentifier((GKInstance) modifiedResidue);
+            if (modification != null)
+                output += modification;
+        }
+        resetIndices();
+
+        for (Object modifiedResidue : modifiedResidues) {
+            modification = residueExporter.exportModificationIdentifier((GKInstance) modifiedResidue);
+            if (modification != null)
+                output += modification;
+        }
+        resetIndices();
+
+        return output;
+    }
+
+    /**
+     * Return the freeText values for a given EWAS.
+     *
+     * @param modifiedResidues
+     * @return String
+     * @throws InvalidAttributeException
+     * @throws Exception
+     */
+    private String getFreeText(List<Object> modifiedResidues) throws InvalidAttributeException, Exception {
+        if (modifiedResidues == null || modifiedResidues.size() == 0)
+            return null;
+        String output = "";
+        AbstractModifiedResidue residueExporter = null;
+        String pkg = this.getClass().getPackage().getName();
+        String cls = null;
+        // Modification classes that have free text columns.
+        List<String> allowedClasses = Arrays.asList(ReactomeJavaConstants.FragmentInsertionModification,
+                                                    ReactomeJavaConstants.FragmentDeletionModification,
+                                                    ReactomeJavaConstants.FragmentReplacedModification,
+                                                    ReactomeJavaConstants.IntraChainCrosslinkedResidue);
+        for (Object modifiedResidue : modifiedResidues) {
+            if (allowedClasses.stream().noneMatch(((GKInstance) modifiedResidue).getSchemClass()::isa))
+                continue;
+
+            if (output.length() > 0)
+                output += ProExporterConstants.freeTextDelimiter;
+
+            cls = ((GKInstance) modifiedResidue).getSchemClass().getName();
+            residueExporter = (AbstractModifiedResidue) Class.forName(pkg + "." + cls).getConstructor().newInstance();
```

Comment:
As the name you used for this variable, residueExporter, your set of AbstractModifiedResidue classes are not really AbstractModifiedResidue class. Names used are confusing.

Is there a need to create an object for each modified residue instance? It seems to me you use the same functions for different instances. Therefore, you need only one copy of object for one type of AbstractModifiedResidue GKInstance.

Reply:
I followed your recommendation and added a "class map" to the ProExporter class. At most, eight instances are created.

```java
Map<SchemaClass, AbstractModifiedResidue> residueClassMap;

public ProExporter() {
    residueClassMap = new HashMap<SchemaClass, AbstractModifiedResidue>();
}
```